packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts

```diff
@@ -0,0 +1,360 @@
+              return aspectExtractedNested;
+            }
+            break;
+          }
+          // Navigation properties
+          case RelationshipMeaning.RelatedInstance: {
+            if (
+              // Deal with a TypeDefinition
+              nestedContentField.pathToPrimaryClass[0].relationshipInfo.name ===
+              "BisCore:GeometricElement3dHasTypeDefinition"
+            ) {
+              return extractNested(["TypeDefinition"], nestedContentField.nestedFields);
+            }
+            break;
+          }
+          default: {
+            // Some elements don't have a path to primary class or relationship meaning..
+            // Most likely a simple struct property
+            if (!nestedContentField.pathToPrimaryClass) {
+              return extractStruct(property);
+            }
+          }
+        }
+      }
+    }
+    return [];
+  });
+
+export const convertToECProperties = (property: PropertyMetaData): ECProperty[] => {
+  const ecProperty: ECProperty = {
+    ecSchemaName: property.schema,
+    ecClassName: property.className,
+    ecPropertyType: property.type,
+    ecPropertyName: "",
+  };
+  switch (property.primitiveNavigationClass) {
+    // Hardcode Models navigation path and label behavior
+    case "BisCore:ModelContainsElements": {
+      return [
+        {
+          ...ecProperty,
+          ecPropertyName: [
+            ...property.propertyTraversal,
+            "ModeledElement",
+            "UserLabel",
+          ].join("."),
+        },
+        {
+          ...ecProperty,
+          ecPropertyName: [
+            ...property.propertyTraversal,
+            "ModeledElement",
+            "CodeValue",
+          ].join("."),
+        },
+      ];
+    }
+    // Hardcode Category and Physical Material navigation path and label behavior
+    case "BisCore:GeometricElement3dIsInCategory":
+    case "BisCore:PhysicalElementIsOfPhysicalMaterial":
+      return [
+        {
+          ...ecProperty,
+          ecPropertyName: [
+            ...property.propertyTraversal,
+            "UserLabel",
+          ].join("."),
+        },
+        {
+          ...ecProperty,
+          ecPropertyName: [
+            ...property.propertyTraversal,
+            "CodeValue",
+          ].join("."),
+        },
+      ];
+    default: {
+      return [
+        {
+          ...ecProperty,
+          ecPropertyName: property.propertyTraversal.join("."),
+        },
+      ];
+    }
+  }
+};
+
+export const findProperties = (ecProperties: ECProperty[], propertiesMetaData: PropertyMetaData[]) => {
+  let ecPropertiesCopy = [...ecProperties];
+  const propertiesMetaDataResult: PropertyMetaData[] = new Array<PropertyMetaData>();
+  let notFound = false;
+  while (ecPropertiesCopy.length !== 0) {
+    for (let i = 0; i < propertiesMetaData.length; i++) {
+      const generatedProperty = convertToECProperties(propertiesMetaData[i]);
```

Comment:
Convert all `propertiesMetaData` with a `map` function once instead of doing the conversion for all of them for each iteration of the outer `while` loop.

Reply:
I will come back to this later as the currently implementation works.